File: cpp/mcts.cpp
#include "mcts.h"

#include "params.h"
#include "network.h"
#include "heuristic_agent.h"

#include <math.h>
#include <cfloat>
#include <functional>
#include <random>

const int BS = BOARD_SIZE;

Network network;
//input_frame_num = 5 means, past 2 moves per each player + 1

HeuristicAgent heuristic;

vector<BoardState> get_next_states(BoardState& state)
{
	vector<BoardState> sl;
	int turn = state.turn + 1;

	for(int i=0; i<BS; i++) {
		for(int j=0; j<BS; j++) {
			if(state[i][j] == 0) {
				BoardState new_state = BoardState(state.board_size, turn, i, j);
				new_state.board = state.board;
				new_state[i][j] = turn;
				sl.push_back(new_state);
			}
		}
	}

	return sl;
}


vector<Board> preproc_board(Board& board, int turn)
{
	int frame = INPUT_FRAME_NUM;
	vector<Board> s(frame, vector<vector<int> > (BS, vector<int>(BS, 0)));
	for(int fn=0; fn<(frame-1)/2; fn++) {
		for(int i=0; i<BS; i++) {
			for(int j=0; j<BS; j++) {
				if(board[i][j] > 0 && board[i][j]%2 == turn%2 && board[i][j] <= turn - fn*2) {
					s[fn][i][j] = 1;
				}
			}
		}
	}
	for(int fn=0; fn<(frame-1)/2; fn++) {
		for(int i=0; i<BS; i++) {
			for(int j=0; j<BS; j++) {
				if(board[i][j] > 0 && board[i][j]%2 != turn%2 && board[i][j] <= turn - fn*2) {
					s[fn + (frame-1)/2][i][j] = 1;
				}
			}
		}
	}

	if(turn%2 == 0) {
		s[frame-1] = vector<vector<int> > (BS, vector<int>(BS, 0));
	} else {
		s[frame-1] = vector<vector<int> > (BS, vector<int>(BS, 1));
	}

	return s;
}

Board dihedral_reflection_rotation(int i, Board x)
{
	function<Board(Board&)> fliplr = [](Board& x)->Board{
		int bs = (int)x.size();
		Board ret = vector<vector<int> > (bs, vector<int>(bs, 0));
		for(int ii=0; ii<bs; ii++) {
			for(int jj=0; jj<bs; jj++) {
				ret[ii][jj] = x[ii][bs-jj-1];
			}
		}
		return ret;
	};
	function<Board(Board&)> rot90 = [](Board& x)->Board{
		int bs = (int)x.size();
		Board ret = vector<vector<int> > (bs, vector<int>(bs, 0));
		for(int ii=0; ii<bs; ii++) {
			for(int jj=0; jj<bs; jj++) {
				ret[bs-jj-1][ii] = x[ii][jj];
			}
		}
		return ret;
	};

	if(i > 0) {
		if(i > 4) {
			x = fliplr(x);
			i -= 4;
		}
		for(int j=0; j<i; j++) {
			x = rot90(x);
		}
	} else if(i < 0) {
		int ii = -i;
		if(ii > 4) {
			ii -= 4;
		}
		for(int j=0; j<4-ii; j++) {
			x = rot90(x);
		}

		if(i < -4) {
			x = fliplr(x);
		}
	}
	return x;
}

vector<vector<double> > dihedral_reflection_rotation(int i, vector<vector<double> > x)
{
	typedef vector<vector<double> > VVD;
	function<VVD(VVD&)> fliplr = [](VVD& x)->VVD{
		int bs = (int)x.size();
		VVD ret = vector<vector<double> > (bs, vector<double>(bs, 0));
		for(int ii=0; ii<bs; ii++) {
			for(int jj=0; jj<bs; jj++) {
				ret[ii][jj] = x[ii][bs-jj-1];
			}
		}
		return ret;
	};
	function<VVD(VVD&)> rot90 = [](VVD& x)->VVD{
		int bs = (int)x.size();
		VVD ret = vector<vector<double> > (bs, vector<double>(bs, 0));
		for(int ii=0; ii<bs; ii++) {
			for(int jj=0; jj<bs; jj++) {
				ret[bs-jj-1][ii] = x[ii][jj];
			}
		}
		return ret;
	};

	if(i > 0) {
		if(i > 4) {
			x = fliplr(x);
			i -= 4;
		}
		for(int j=0; j<i; j++) {
			x = rot90(x);
		}
	} else if(i < 0) {
		int ii = -i;
		if(ii > 4) {
			ii -= 4;
		}
		for(int j=0; j<4-ii; j++) {
			x = rot90(x);
		}

		if(i < -4) {
			x = fliplr(x);
		}
	}
	return x;
}




PV_pair evaluate_with_network(BoardState& state, vector<BoardState>& state_list)
{
	int i = rand()%8 + 1;
	Board board = dihedral_reflection_rotation(i, state.board);
	int turn = state.turn;
	vector<Board> s = preproc_board(board, turn);

	pair<vector<vector<double> >, double> pv = network.get_output(s);
	auto p = pv.first;
	double v = pv.second;
	p = dihedral_reflection_rotation(-i, p);
	map<BoardState, double> p_dict;
	for(auto& new_state : state_list) {
		int r = new_state.last_row;
		int c = new_state.last_col;
		p_dict[new_state] = p[r][c];
	}
	return make_pair(p_dict, v);
}



PV_pair evaluate_with_random(BoardState& state, vector<BoardState>& state_list)
{
	double v = (rand()%2000)/1000.0 - 1;
	map<BoardState, double> p_dict;
	double p_sum = 0;
	for(auto& new_state : state_list) {
		int r = new_state.last_row;
		int c = new_state.last_col;
		double p = (rand()%1000)/1000.0;
		p_dict[new_state] = p;
		p_sum += p;
	}

	for(auto& p : p_dict) {
		p.second /= p_sum;
	}

	return make_pair(p_dict, v);
}



PV_pair evaluate_with_heuristic(BoardState& state, vector<BoardState>& state_list)
{
	return heuristic.evaluate(state, state_list);
}



Node::Node(
	BoardState s, 
	PV_pair (*f) (BoardState& board_state, vector<BoardState>& state_list), 
	Node* p
)
{
	evaluate = f;
	parent = p;
	state = s;
	N = 0;
	N_sum = 0;
	Q = 0;
	W = 0;
	P = 0;
	U = 0;
	selected_child = NULL;
}

void Node::search()
{
	Node* max_child = select();
	if(max_child == this) {
		max_child->expand();
	} else {
		max_child->search();
	}
}

vector<double> Dirichlet(double alpha, int size)
{
	default_random_engine generator;
	gamma_distribution<double> distribution(alpha, 1);
	vector<double> noise = vector<double>(size, 0);
	double sum = 0;
	for(int i=0; i<size; i++) {
		noise[i] = distribution(generator);
		sum += noise[i];
	}
	for(int i=0; i<size; i++) {
		noise[i] /= sum;
	}

	return noise;
}

Node* Node::select()
{
	double p, u;
	Node* max_child = this;
	
	if((int)child_list.size() == 0)
		return max_child;

	double QU_max = -DBL_MIN;
	const double constant = C_PUCT * sqrt(N_sum);
	vector<double> noise = Dirichlet(DIR_ALPHA, (int)child_list.size());
	int idx = 0;
	for(auto& child : child_list) {
		if(parent == NULL) {
			p = (1-EPSILON)*(child->P) + EPSILON*noise[idx++];
		} else {
			p = (child->P);
		}
		u = p * constant / (1 + (child->N));
		child->U = u;
		double QU = child->Q + u;
		if(QU_max < QU) {
			QU_max = QU;
			max_child = child;
		}
	}
	return max_child;
}

void Node::expand()
{
	if(is_game_ended(state.board)) {
		vector<BoardState> emp(0);
		double v = evaluate(state, emp).second;
		if(parent != NULL) {
			parent->backup(v, this);
		}
		return;
	}

	vector<BoardState> state_list = get_next_states(state);
	PV_pair pv = evaluate(state, state_list);
	Dict p = pv.first;
	double v = pv.second;
	for(auto& state : state_list) {
		Node* new_node = new Node(state, evaluate, this);
		child_list.push_back(new_node);
		new_node->P = p[state];
	}
	if(parent != NULL) {
		parent->backup(v, this);
	}	
}

void Node::backup(double v, Node* child)
{
	child->N += 1;
	N_sum++;
	child->W += v;
	child->Q = child->W / child->N;
	if(parent != NULL) {
		parent->backup(v, this);
	}
}

pair<map<Node*, long double>, long double> Node::get_pi(double t)
{
	map<Node*, long double> pi;
	long double N_s = 1e-9;
	for(auto& n : child_list) {
		pi[n] = powl((long double)n->N, (long double)1.0/t);
		N_s += pi[n];
	}

	for(auto& n : pi) {
		n.second /= N_s;
	}

	return make_pair(pi, N_s);
}

Node* Node::play(double t)
{
	pair<map<Node*, long double>, long double> pi_n = get_pi(t);
	map<Node*, long double> pi = pi_n.first;
	long double N_sum = pi_n.second;
	vector<pair<long double, Node*> > N_list;
	for(auto& n : pi) {
		N_list.push_back(make_pair(n.second, n.first));
	}
	for(int i=1; i<(int)N_list.size(); i++) {
		N_list[i].first += N_list[i-1].first;
	}
	long double r = rand() / (long double)RAND_MAX;
	for(int i=0; i<(int)N_list.size(); i++) {
		if(N_list[i].first >= r) {
			return N_list[i].second;
		}
	}

	printf("Exception on \'play\' %d\n", (int)child_list.size());
	printf("Exception on \'play\' %d\n", (int)N_list.size());
	printf("Exception on \'play\' %d\n", (int)pi.size());

	return N_list[-1].second;
}

void Node::clear()
{
	for(auto& child : child_list) {
		child->clear();
	}
	delete(this);
}



MCTS::MCTS(
	int board_size, 
	PV_pair (*evaluate) (BoardState& board_state, vector<BoardState>& state_list)
)
{
	root = new Node(BoardState(board_size, 0), evaluate);
	//network = Network(BS, 5, 9, false); //FIXME
}

void MCTS::search(Node* node)
{
	if(node==NULL) {
		root->search();
	} else {
		node->search();
	}
}

Node* MCTS::play(Node* node, double t)
{
	Node* next_node = node->play(t);
	node->selected_child = next_node;
	return next_node;
}

File: python/MCTS.py
from params import *
from Gomoku import *
from Network import Network
from HeuristicAgent import HeuristicAgent
import random
import math

BS = BOARD_SIZE

heuristic = HeuristicAgent()

network = Network(board_size = BS, input_frame_num = INPUT_FRAME_NUM, residual_num = RESIDUAL_NUM, is_trainable=True) #False
# input_frame_num = 5 means, past 2 mover per each player + 1

def get_next_states(state):
    sl = []
    turn = state.turn + 1

    for i in range(BS):
        for j in range(BS):
            if state[i][j] == 0:
                new_state = BoardState(state.board_size, turn, i, j)
                new_state.board = np.copy(state.board)
                new_state[i][j] = turn
                sl.append(new_state)
    return sl

def preproc_board(board, turn):
    frame = INPUT_FRAME_NUM
    turn_parity = turn % 2
    half_frame = (frame-1)//2
    
    s = np.zeros([1, BS, BS, frame])
    for i in range(BS):
        for j in range(BS):
            if 0 == board[i][j]:
                continue
            for fn in range(half_frame):
                limit = turn - fn*2
                if board[i][j] <= limit:
                    if board[i][j] % 2 == turn_parity:
                        s[0][i][j][fn] = 1
                    else:
                        s[0][i][j][fn+half_frame] = 1
        
    if turn_parity == 1:
        s[0,:,:,-1] = np.ones([BS, BS]) 
    
    return s

def dihedral_reflection_rotation(i, x):
    if i > 0:
        if i > 4:
            x = np.fliplr(x)
            i -= 4
        for _ in range(i):
            x = np.rot90(x)
    elif i < 0:
        ii = -i
        if ii > 4:
            ii -= 4
        for _ in range(4 - ii):
            x = np.rot90(x)

        if i < -4:
            x = np.fliplr(x)
    return x

def evaluate_with_network(state, state_list):
    i = random.randrange(1, 9)
    board = dihedral_reflection_rotation(i, np.asarray(state.board))
    turn = state.turn
    s = preproc_board(board, turn)
    p,v = network.get_output(s)
    p = np.reshape(p, (BS, BS))
    p = dihedral_reflection_rotation(-i, p)
    p_dict = {}
    for new_state in state_list:
        r = new_state.last_row
        c = new_state.last_col
        p_dict[new_state] = p[r][c]

    if is_game_ended(state.board):
        v = 1

    return p_dict,v

def evaluate_with_constant(state, state_list):
    if is_game_ended(state.board):
        v = state.turn%2==0 #1 when mcts starts first, 0 when mcts starts later
    else:
        v = 0.5
    
    p_dict = {}
    p_sum = 0
    for new_state in state_list:
        r = new_state.last_row
        c = new_state.last_col
        p_dict[new_state] = 1
        p_sum += 1

    for p in p_dict:
        p_dict[p] /= p_sum

    return p_dict,v


def evaluate_with_random(state, state_list):
    v = random.uniform(-1, 1)
    p_dict = {}
    p_sum = 0
    for new_state in state_list:
        r = new_state.last_row
        c = new_state.last_col
        p = random.uniform(0, 1)
        p_dict[new_state] = p
        p_sum += p

    for p in p_dict:
        p_dict[p] /= p_sum

    return p_dict,v


def evaluate_with_heuristic(state, state_list):
    p_dict, v = heuristic.evaluate(state, state_list)
    return p_dict, v

class Node:
    def __init__(self, state, evaluate, parent=None):
        self.evaluate = evaluate
        self.parent = parent
        self.state = state
        self.child_list = []
        self.N = 0
        self.N_sum = 0
        self.Q = 0
        self.W = 0
        self.P = 0
        self.U = 0
        self.selected_child = None
    
    def search(self):
        max_child = self.select()
        if max_child == self:
            max_child.expand()
        else:
            #self.selected_child = max_child
            max_child.search()
    
    def select(self):
        N_sum = self.N_sum
        COMP_EPSILON = 1-EPSILON
        child_list = self.child_list
        child_num = len(child_list)
        
        if child_num == 0:
            return self

        const = C_PUCT * (N_sum**0.5)
        
        if self.parent == None:
            noise = np.random.dirichlet([DIR_ALPHA] * child_num)
            idx = 0
            for child in child_list:
                P = COMP_EPSILON*child.P + EPSILON*noise[idx]
                idx += 1
                
                child.U = P * const / (1 + child.N)
        else:
            for child in child_list:
                child.U = child.P * const / (1 + child.N)

        return max(child_list, key=lambda x: x.Q + x.U)

    def expand(self):
        if is_game_ended(self.state.board): # FIXME Gomoku specific
            _,v = self.evaluate(self.state, [])
            if self.parent != None:
                self.parent.backup(v, self)
            return

        state_list = get_next_states(self.state)
        p,v = self.evaluate(self.state, state_list)
        for state in state_list:
            new_node = Node(state, self.evaluate, self)
            self.child_list.append(new_node)
            new_node.P = p[state]
        if self.parent != None:
            self.parent.backup(v, self)

    def backup(self, v, child):
        child.N += 1
        self.N_sum += 1
        child.W+= v
        child.Q = child.W / child.N
        if self.parent != None:
            self.parent.backup(-v, self)
    
    def get_pi(self, t):
        pi = {}
        N_s = 1e-9
        if t == 0:
            max_n = 0
            for n in self.child_list:
                if max_n < n.N:
                    max_n = n.N
                    max_child = n
            for n in self.child_list:
                if n == max_child:
                    pi[n] = 1
                else:
                    pi[n] = 0
            N_s = 1
        else:
            for n in self.child_list:
                pi[n] = n.N ** (1/t)
                N_s += pi[n]
            
            for n in pi:
                pi[n] /= N_s

        return pi, N_s

    def play(self, t):
        pi, N_sum = self.get_pi(t)
        N_list = []
        for n in pi:
            N_list.append([pi[n],n])
        for i in range(1, len(N_list)):
            N_list[i][0] += N_list[i-1][0]
        r = np.random.uniform(0, 1)
        for i in range(len(N_list)):
            if N_list[i][0] >= r:
                return N_list[i][1]
        
        print('Exception on \'play\' line:196 ', len(self.child_list))
        print('Exception on \'play\' line:197 ', len(N_list))
        print('Exception on \'play\' line:198 ', len(pi))
        return N_list[-1][1]
    
    
class MCTS:
    def __init__(self, board_size, evaluate):
        self.root = Node(BoardState(board_size, 0), evaluate)

    def search(self, node=None):
        if node == None:
            self.root.search()
        else:
            node.search()

    def play(self, node, t):
        next_node = node.play(t)
        node.selected_child = next_node
        return next_node


